### PR TITLE
[BACK-69] Improving Categories and making them more specific

### PR DESCRIPTION
**Pull request by acreturs**
### Description of Issue ( new by @thomas-woehrle )

Compared to the first version there should be less categories. They are too fine-granular at the moment. Try to cut it down to 4-6 categories.
Prompts of first version were not ideal.

Overall there were several improvements to be made to the categorization method to be fully functional

Vague example of what i mean: https://chat.openai.com/share/48bd44b4-13a7-4ff8-9938-2214f0b17f6b
### How I implemented ( new by @thomas-woehrle  )

- adjusted getRepositoryTopics to return a list of strings rather than a string
- input to getCategoryByNumber is a list of topics and a description, but can be null
- call to GPT to get a number/multiple numbers for a project
- I use numbers instead of strings because this does leave no room for GPT to make up stuff like new category names
- transforms GPTs answer into a list of Categories

### CODE REVIEW

File: packages/graphql_server/src/api/openAIApi.ts [hunk 1]
```diff
@@ -88,7 +88,7 @@ async function getHackernewsSentiment(comments: string) {
     }
   } catch (error) {
     console.log('OpenAI request did not work out: ', error)
-    return null
+    return null // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (DogukanGun) [inline packages/graphql_server/src/api/openAIApi.ts:91]: Please do not disable eslint feature in this line

File: packages/graphql_server/src/api/openAIApi.ts [hunk 2]
```diff
@@ -98,60 +98,80 @@ async function getHackernewsSentiment(comments: string) {
...
-  ]
+  // ML/AI
+  MachineLearning = 'Machine Learning', // ML/AI
+  NaturalLanguageProcessing = 'Natural Language Processing', // ML/AI
+  ComputerVision = 'Computer Vision' // ML/AI
+}
 
-  const questionCategoriesSpecific = //this is the question send to openAI
-    'These categories should be used to categorize a software engineering project. Please choose two of the categories from the list that describe this project the best based on these words or readme (Your response should only consist of the two words you choose, separated by a comma): '
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (DogukanGun) [inline packages/graphql_server/src/api/openAIApi.ts:123]: Same topic

**Review by DogukanGun — DISMISSED** [review]

**Review by thomas-woehrle — DISMISSED** [review]
Have to make this review like this because there are some problems with commenting where I want to. PR goes into a good direction.

Not comprehensive list of things that are not perfect:
- `categorizeProjectGeneral()`  does not categorize the projects in general but in detail. 
- when testing it for repo gpt-engineer of AntonOsika it returns a odd category:
Screenshot 2023-06-17 at 14 09 19
- test it for a few repos and see if it works
- It returns a string and not a list of categories.

**Comment by thomas-woehrle** [comment]
Also texted you on slack @acreturs

**Comment by thomas-woehrle** [comment]
Not final yet, but I only plan to change minor things. For example return type of getCategoryNumber to number and chnage the categories. 
Because we are short on time, I will set it to _ready for review_ still

File: packages/graphql_server/src/api/openAIApi.ts [hunk 3]
```diff
@@ -11,7 +12,7 @@ const errorMessage =
 //The Header for all requests
 const headers = {
   'Content-Type': 'application/json',
-  Authorization: 'Bearer ' + process.env.OPENAI_API_KEY
+  Authorization: 'Bearer ' + process.env.OPENAI_API_KEY //process.env.OPENAI_API_KEY
```
Comment (pczern) [inline packages/graphql_server/src/api/openAIApi.ts:15]: remove comment

File: packages/graphql_server/src/api/openAIApi.ts [hunk 4]
```diff
@@ -45,6 +46,7 @@ async function getELI5FromReadMe(readMe: string) {
       return null
     } else {
       const content: string = data.choices[0].message.content
+      console.log(content)
```
Comment (pczern) [inline packages/graphql_server/src/api/openAIApi.ts:49]: remove log

File: packages/graphql_server/src/api/openAIApi.ts [hunk 5]
```diff
@@ -92,151 +94,52 @@ async function getHackernewsSentiment(comments: string) {
...
-    case 4: //if chatgtp says the categories are not specific enough we need t use the readme file to put it into two of the categories
-      categoriesSpecific = listOfCategoriesDeveloperTools.concat(listOfCategoriesInfrastructure)
-      categoriesSpecific = categoriesSpecific.concat(listOfCategoriesMLAI)
-      break
-    default:
-      console.log('Invalid categoryGeneral value')
-      return null
+const createCategorizationPrompt = () => {
+  let prompt = 'Answer with '
+  console.log(Object.entries(topicsMap))
```
Comment (pczern) [inline packages/graphql_server/src/api/openAIApi.ts:107]: remove log

File: packages/graphql_server/src/api/openAIApi.ts [hunk 6]
```diff
@@ -92,151 +94,52 @@ async function getHackernewsSentiment(comments: string) {
...
+  for (const [key, value] of Object.entries(topicsMap)) {
+    if (key === '9') continue
```
Comment (pczern) [inline packages/graphql_server/src/api/openAIApi.ts:109]: typescript supports enums btw

File: packages/graphql_server/src/api/githubApi.ts [hunk 7]
```diff
@@ -179,24 +179,20 @@ export async function getRepositoryTopics(
...
-      return topics.join(' ') //return the openai response as a string
-    } else {
-      throw new Error('No repository topics found.')
-    }
-  } catch (error) {
-    console.log('Could not retrieve the categories')
-    return ' '
+    return topics
+  } else {
+    return [] as string[]
```
Comment (pczern) [inline packages/graphql_server/src/api/githubApi.ts:195]: it's best to return null

File: packages/graphql_server/src/api/openAIApi.ts [hunk 8]
```diff
@@ -92,151 +94,52 @@ async function getHackernewsSentiment(comments: string) {
...
+        You take in infos about a repository that is hosted on github and reply with numbers separated by commas to categorize the project.
+        Your answer may only contain numbers separated by commas.
+        ${createCategorizationPrompt()}
+        `
       },
       {
         role: 'user',
-        content: categoriesSpecific.join(' ,') + questionCategoriesSpecific + readMeOrCategory
+        content: `        
+        Topics related to the repository: ${categories.toString()}
```
Comment (pczern) [inline packages/graphql_server/src/api/openAIApi.ts:130]: why does this string span lines?

File: packages/graphql_server/src/api/openAIApi.ts (hunk 6 above)
Comment (acreturs) [reply packages/graphql_server/src/api/openAIApi.ts:109]: not sure what you mean

**Review by thomas-woehrle — APPROVED** [review]

**Review by DogukanGun — APPROVED** [review]
There is a warning, it may be fixed later.
